### PR TITLE
libslirp: fix CVE-2020-29129

### PR DESCRIPTION
**Pull request by mweinelt**
###### Motivation for this change

Fixes out-of-bounds access while processing ARP/NCSI packets.

Fixes: CVE-2020-29129

https://lists.freedesktop.org/archives/slirp/2020-November/000115.html

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by stigtsp — COMMENTED** [review]
Result of `nixpkgs-review pr 105157` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>buildah</li>
    <li>libslirp</li>
    <li>podman</li>
    <li>slirp4netns</li>
  </ul>
</details>

**Comment by mweinelt** [comment]
FWIW: QEMU will still be vulnerable, because we are currently using a vendored libslirp version. 

#101608 was going to change that, but didn't happen. Can I pick up those changes? @zowoq

File: pkgs/development/libraries/libslirp/default.nix [hunk 1]
```diff
@@ -18,6 +18,12 @@ stdenv.mkDerivation rec {
     sha256 = "0pzgjj2x2vrjshrzrl2x39xp5lgwg4b4y9vs8xvadh1ycl10v3fv";
   };
 
+  patches = [
+    # https://lists.freedesktop.org/archives/slirp/2020-November/000115.html
+    # remove >4.3.1
+    ./CVE-2020-29129.patch
```
Comment (zowoq) [inline pkgs/development/libraries/libslirp/default.nix:24]: Could use `fetchpatch`?

https://gitlab.freedesktop.org/slirp/libslirp/-/commit/2e1dcbc0c2af64fcb17009eaf2ceedd81be2b27f.patch

Comment (mweinelt) [reply pkgs/development/libraries/libslirp/default.nix]: Done.

File: pkgs/development/libraries/libslirp/CVE-2020-29129.patch [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ 
++    if (pkt_len < ETH_HLEN + sizeof(struct slirp_arphdr)) {
++        return; /* packet too short */
++    }
++
+     ar_op = ntohs(ah->ar_op);
+     switch (ar_op) {
+     case ARPOP_REQUEST:
+-- 
+2.28.0
```
Comment (zowoq) [inline pkgs/development/libraries/libslirp/CVE-2020-29129.patch:46]: The patch file can be removed.

**Comment by mweinelt** [comment]
Patched the vendored libslirp in QEMU as well.

**Comment by mweinelt** [comment]
Ported to 20.09 in b7fd0c05a61bd78d9d98660714ebda26c9e36d74 aa8ba404e2af68a4b84de5b714a9951fa5664290

**Pull request merged**